lightning/src/routing/router.rs

```diff
@@ -1564,45 +1564,61 @@ fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters,
+		let third_last_id = NodeId::from_pubkey(&path.get(path.len().saturating_sub(3)).unwrap().pubkey);
+		let mut nodes_to_avoid: [NodeId; 3] = [third_last_id, second_last_id, last_id];
+		let mut nodes_ring_index = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk.
+		let mut cur_hop: Option<NodeId> = None;
+		let mut path_nonce = [0u8; 12];
+		if let Some(starting_hop) = path.iter().rev()
+			.find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
+				cur_hop = Some(NodeId::from_pubkey(&starting_hop.pubkey));
+				path_nonce.copy_from_slice(&cur_hop.unwrap().as_slice()[..12]);
+		}
+
+		// Init PRNG with the path-dependant nonce, which is static for private paths.
+		let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+		let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
 
-			// Init PRNG with path nonce
-			let mut path_nonce = [0u8; 12];
-			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..12]);
-			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
-			let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
+		// Pick a random path length in [1 .. 3]
+		prng.process_in_place(&mut random_path_bytes);
+		let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
 
-			// Pick a random path length in [1 .. 3]
-			prng.process_in_place(&mut random_path_bytes);
-			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+		for _random_hop in 0..random_walk_length {
+			// If we don't find a suitable offset in the public network graph, we default to
+			// MEDIAN_HOP_CLTV_EXPIRY_DELTA.
+			let mut random_hop_offset = MEDIAN_HOP_CLTV_EXPIRY_DELTA;
 
-			for _random_hop in 0..random_walk_length {
+			if let Some(cur_node_id) = cur_hop {
 				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
-					// Randomly choose the next hop
+					// Randomly choose the next unvisited hop.
 					prng.process_in_place(&mut random_path_bytes);
-					if let Some(random_channel) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len())
+					if let Some(random_channel) = usize::from_be_bytes(random_path_bytes)
+						.checked_rem(cur_node.channels.len())
 						.and_then(|index| cur_node.channels.get(index))
 						.and_then(|id| network_channels.get(id)) {
 							random_channel.as_directed_from(&cur_node_id).map(|(dir_info, next_id)| {
-								dir_info.direction().map(|channel_update_info|
-									shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
-										.checked_add(channel_update_info.cltv_expiry_delta.into())
-										.unwrap_or(shadow_ctlv_expiry_delta_offset));
-								cur_node_id = *next_id;
+								if !nodes_to_avoid.iter().any(|x| x == next_id) {
+									nodes_to_avoid[nodes_ring_index] = *next_id;
+									nodes_ring_index = (nodes_ring_index + 1) % 3;
```

Comment:
Regarding the modulo operator and name `nodes_ring_index`, is it ever the case we'll loop back to the `0` index? Seems like we'll always loop use at most `3` times.

Reply:
Ah good point. In https://github.com/lightningdevkit/rust-lightning/pull/1360/commits/eb5a4c478aa2c98d0c04f38615e66296c673c20b I now simply use the previously unused `random_hop` index value of the `for` loop.